## Insulin drug data analysis

### Gather

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

In [3]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [4]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [5]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [6]:
treatments.count()

given_name      280
surname         280
auralin         280
novodra         280
hba1c_start     280
hba1c_end       280
hba1c_change    171
dtype: int64

In [7]:
adverse_reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [8]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


### Assessment - quality issue
patients table: 
* zip code data type incorrect due to string float conversion
* zip code padding zero issue, wrong number of digits
* wrong height data e.g. 27in, data anomaly
* inconsistent State data e.g.fullname vs abbv.
* name typo,  given_name for the patient with the patient_id 9, Dsvid Gustafsson
* Missing demographic information ('address' to 'contact' columns)with null values, missing entries null values
* Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
* Multiple phone number formats
* Default John Doe data
* Multiple records for Jakobsen, Gersten, Taylor
* kgs instead of lbs for Zaitseva weight

treatments table:
* missing value NaN in HbA1c change
* letter 'u' next to the start dose and end dose in the auralin and novodra columns, data type should be integers
* lowercase given name & surname
* missing record count(only 280 instead of 350)
* hba1c_change column missing record(only 171)
* Erroneous datatypes (auralin and novodra columns)
* Inaccurate HbA1c changes (4s mistaken as 9s) inaccurate value
* Nulls represented as dashes (-) in auralin and novodra columns

adverse_reactions table:
* lowercase for given name & surname

In [9]:
treatments.sample(5)

given_name   surname    auralin    novodra  hba1c_start  hba1c_end  \
188      bouke    glaser          -  26u - 30u         7.72       7.34   
166      annie     allen  36u - 42u          -         9.95       9.58   
189        mee     chung  42u - 56u          -         7.56       7.20   
42      marija  grubišić  37u - 43u          -         7.53       7.15   
107   flaviano  endrizzi          -  46u - 43u         7.93       7.56   

     hba1c_change  
188           NaN  
166          0.37  
189          0.36  
42           0.38  
107          0.37

In [10]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [11]:
# show reaction values count
adverse_reactions.adverse_reaction.value_counts()['cough']

2

In [12]:
patients['city'].value_counts()

New York         18
San Diego         8
Tulsa             7
Atlanta           6
Houston           6
                 ..
Yazoo City        1
Olathe            1
Winnsboro         1
Chester           1
Overland Park     1
Name: city, Length: 349, dtype: int64

In [13]:
patients.address.value_counts()

123 Main Street             6
2778 North Avenue           2
2476 Fulton Street          2
648 Old Dear Lane           2
3094 Oral Lake Road         1
                           ..
1066 Goosetown Drive        1
4291 Patton Lane            1
4643 Reeves Street          1
174 Lost Creek Road         1
3652 Boone Crockett Lane    1
Name: address, Length: 483, dtype: int64

In [14]:
# check count of specific city of patient
patients[patients['city']=='San Diego'].city.count()

8

In [15]:
patients.count()

patient_id      503
assigned_sex    503
given_name      503
surname         503
address         491
city            491
state           491
zip_code        491
country         491
contact         491
birthdate       503
weight          503
height          503
bmi             503
dtype: int64

In [16]:
# check null value in address field
patients[patients['address'].isnull()].sum()['address']

0

In [17]:
# check null value in contact field
patients['contact'].isnull().sum()

12

In [18]:
# check duplicates
patients[patients.address.duplicated()].count()

patient_id      19
assigned_sex    19
given_name      19
surname         19
address          8
city             8
state            8
zip_code         8
country          8
contact          8
birthdate       19
weight          19
height          19
bmi             19
dtype: int64

In [19]:
# check duplicates name
patients[patients.given_name.duplicated()].given_name

69        Joseph
72         Søren
104       Yasmin
109      Stephen
115     Caroline
153         John
188        Søren
215         John
229         John
230    Elisabeth
237         John
242         John
244         John
251         John
262      Rebecca
277         John
282        Sandy
325         Sara
332         Abel
341      Fatimah
365       Robert
375        Sofia
386      Sargent
394         Anja
403       Robert
417       Xiuxiu
433          Leo
440        David
470     Muhammad
472        Julia
481      Michael
496    Alexander
498      Mustafa
Name: given_name, dtype: object

In [20]:
patients.weight.sort_values()

210     48.8
459    102.1
335    102.7
74     103.2
317    106.0
       ...  
144    244.9
61     244.9
283    245.5
118    254.5
485    255.9
Name: weight, Length: 503, dtype: float64

In [21]:
sum(treatments.auralin.isnull())

0

In [22]:
sum(treatments.novodra.isnull())

0

In [23]:
treatments.auralin.value_counts()

-            143
31u - 38u      3
29u - 37u      3
32u - 42u      2
42u - 49u      2
            ... 
53u - 64u      1
25u - 31u      1
39u - 45u      1
31u - 45u      1
53u - 56u      1
Name: auralin, Length: 121, dtype: int64

In [24]:
# count number of records where patient from new york
len(patients.loc[patients['city'] == 'New York'])

18

In [25]:
# duplicated surname
patients[patients['surname']=='Doe'].surname.count()

6

In [26]:
# weight of the record with data validity issue
weight_lbs = patients[patients.surname == 'Zaitseva'].weight * 2.20462
height_in = patients[patients.surname == 'Zaitseva'].height
bmi_check = 703 * weight_lbs / ( height_in * height_in )
print(bmi_check)

210    19.055827
dtype: float64


In [27]:
patients[patients.surname == 'Zaitseva'].bmi

210    19.1
Name: bmi, dtype: float64

### Tidiness issue

* Contact column in patients table should be split into phone number & email address
* variables in columns of Treatments of table - treatment(auralin & novodra), start dose, end dose
* Adverse reaction join with treatments table
* Given name and surname columns in patients table duplicated in treatments and adverse_reactions tables

In [28]:
# check duplicated columns across tables as series,combine patients and treatments tables for patient info
all_columns = pd.Series(list(patients)+list(treatments)+list(adverse_reactions))
all_columns[all_columns.duplicated()==True]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

In [29]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


### Cleaning

In [38]:
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

In [32]:
# Define: Convert zip code column's data type from float to string type
# convert to string & zero padding to 5 char
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5,fillchar = '0')
patients_clean.zip_code.head()

0    92390
1    61812
2    68467
3    07095
4    36303
Name: zip_code, dtype: object

In [37]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [35]:
patients_clean.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state zip_code        country  \
0  Rancho California  California    92390  United States   
1          Armstrong    Illinois    61812  United States   
2               York    Nebraska    68467  United States   
3         Woodbridge          NJ    07095  United States   
4             Dothan          AL    36303  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [42]:
# issue: missing data
# Define: Treatments table Missing records (280 instead of 350)
# concatenate treatments_cut with the original treatments DataFrame
treatments_cut = pd.read_csv('treatments_cut.csv')
treatments_cut.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38

In [50]:
treatments_clean = pd.concat([treatments,treatments_cut],ignore_index=True)
treatments_clean.count()

given_name      350
surname         350
auralin         350
novodra         350
hba1c_start     350
hba1c_end       350
hba1c_change    213
dtype: int64

In [54]:
# treatments: Missing HbA1c changes and inaccurate HbA1c changes (leading 4s mistaken as 9s)
treatments.hba1c_change.value_counts()

0.39    17
0.35    12
0.33    12
0.38    11
0.37    11
0.36    10
0.92    10
0.34     9
0.97     8
0.93     7
0.90     6
0.30     6
0.32     6
0.95     5
0.96     5
0.31     5
0.91     5
0.94     4
0.51     3
0.27     3
0.29     3
0.99     3
0.98     3
0.54     1
0.28     1
0.24     1
0.53     1
0.26     1
0.50     1
0.20     1
Name: hba1c_change, dtype: int64

In [55]:
# define: Recalculate the hba1c_change column: hba1c_start minus hba1c_end
treatments_clean.hba1c_change = (treatments_clean.hba1c_start - 
                                 treatments_clean.hba1c_end)

In [64]:
treatments_clean[treatments_clean.hba1c_change.isnull()==True]

Empty DataFrame
Columns: [given_name, surname, auralin, novodra, hba1c_start, hba1c_end, hba1c_change]
Index: []

In [69]:
treatments_clean.hba1c_change.count()

350